test/units/parsing/test_ajson.py

```diff
@@ -20,3 +25,159 @@ def test_AnsibleJSONDecoder_vault():
     assert isinstance(data['password'], AnsibleVaultEncryptedUnicode)
     assert isinstance(data['bar']['baz'][0]['password'], AnsibleVaultEncryptedUnicode)
     assert isinstance(data['foo']['password'], AnsibleVaultEncryptedUnicode)
+
+
+class TestAnsibleJSONEncoder():
+    """
+    Class for testing AnsibleJSONEncoder
+    """
+
+    #################
+    # Fixture section
+
+    @pytest.fixture
+    def mapping(self, request):
+        """
+        Returns object of Mapping mock class
+        """
+        class M(Mapping):
+            def __init__(self, *args, **kwargs):
+                self.__dict__.update(*args, **kwargs)
+
+            def __getitem__(self, key):
+                return self.__dict__[key]
+
+            def __iter__(self):
+                return iter(self.__dict__)
+
+            def __len__(self):
+                return len(self.__dict__)
+
+        return M(request.param)
+
+    @pytest.fixture
+    def ansible_json_encoder(self):
+        return AnsibleJSONEncoder()
+
+    @pytest.fixture
+    def vault_data(self):
+        with open(os.path.join(os.path.dirname(__file__), 'fixtures/ajson.json')) as f:
+            data = json.load(f, cls=AnsibleJSONDecoder)
+        return (data['password'], data['bar']['baz'][0]['password'])
+
+    # The end of the fixture section
+    ################################
+
+    @pytest.mark.parametrize(
+        'test_input,expected',
+        [
+            (datetime(2019, 5, 14, 13, 39, 38, 569047), '2019-05-14T13:39:38.569047'),
+            (datetime(2019, 5, 14, 13, 47, 16, 923866), '2019-05-14T13:47:16.923866'),
+            (date(2019, 5, 14), '2019-05-14'),
+            (date(2020, 5, 14), '2020-05-14'),
+        ]
+    )
+    def test_date_datetime(self, ansible_json_encoder, test_input, expected):
+        """
+        Test for passing datetime.date or datetime.datetime objects
+        to AnsibleJSONEncoder.default()
+        """
+        assert(ansible_json_encoder.default(test_input) == expected)
+
+    @pytest.mark.parametrize(
+        'mapping,expected',
+        [
+            ({1: 1}, {1: 1}),
+            ({2: 2}, {2: 2}),
+            ({1: 2}, {1: 2}),
+            ({2: 1}, {2: 1}),
+        ], indirect=['mapping']
+    )
+    def test_mapping(self, ansible_json_encoder, mapping, expected):
+        """
+        Test for passing Mapping object to AnsibleJSONEncoder.default()
+        """
+        m = mapping
+        assert(isinstance(m, Mapping))
+        assert(ansible_json_encoder.default(m) == expected)
+
+    # The pytest.mark.parametrize decorator is not used for this function
+    # because, at least, it's difficult to read it in this particular case.
```

Comment:
If you're concerned about the readability of having the data inline in the parametrize decorator, why not put it in a variable and reference that from the decorator instead?

Reply:
@mattclay , I moved it to auxiliary function, PTAL